tests/components/esphome/test_config_flow.py

```diff
@@ -0,0 +1,179 @@
+"""Test config flow."""
+from collections import namedtuple
+from unittest.mock import patch, MagicMock
+
+import pytest
+
+from homeassistant.components.esphome import config_flow
+from tests.common import mock_coro
+
+MockDeviceInfo = namedtuple("DeviceInfo", ["uses_password", "name"])
+
+
+@pytest.fixture(autouse=True)
+def aioesphomeapi_mock():
+    """Mock aioesphomeapi."""
+    with patch.dict('sys.modules', {
+        'aioesphomeapi': MagicMock(),
+    }):
+        yield
+
+
+@pytest.fixture(autouse=True)
+def mock_api_connection_error():
+    """Mock out the try login method."""
+    with patch('aioesphomeapi.APIConnectionError',
+               new_callable=lambda: OSError) as mock_error:
+        yield mock_error
+
+
+async def test_user_connection_works(hass):
+    """Test we can finish a config flow."""
+    flow = config_flow.EsphomeFlowHandler()
+    flow.hass = hass
+    result = await flow.async_step_user(user_input=None)
+    assert result['type'] == 'form'
+
+    with patch('aioesphomeapi.APIClient') as mock_client:
```

Comment:
Try to break out these patches, that are repeated in the tests with only slight changes, into one or more pytest fixtures. That will make it more DRY. See tellduslive and daikin config flow tests for examples. It's possible to overwrite pytest fixtures per test if needed.